lukaszgardecki/library-app
# [FEATURE][F-60] Internationalization

## 🚀 What?
This pull request introduces internationalization (i18n) to the application, enabling users to interact with the app in their preferred language. The changes span both the frontend and backend to ensure a seamless multilingual experience.

### Frontend:
- Installed `@ngx-translate` and `@angular/localize` in Angular and added `TranslateModule` to `app.module.ts`.
- Updated `StorageService` to provide methods for saving and retrieving the user's preferred language from local storage.
- Created `.json` translation files for English, Polish, Ukrainian, German, Spanish, Czech, and Slovak.
- Added a dropdown to the home page for language selection.
- Created a global `TEXT` object containing references to translation paths.
- Translated all pages and texts in the application.
- Updated the interceptor to include the `Accept-Language` header with the user's preferred language in each outgoing request.

### Backend:
- Converted the `Message` class into an enum, with each enum containing a `.properties` key for translation, enhancing readability.
- Created `.properties` translation files for English, Polish, Ukrainian, German, Spanish, Czech, and Slovak.

## 🎯 Why?
To provide a better user experience by allowing users to interact with the application in their preferred language, making the app more accessible and user-friendly for a diverse audience.

## 🛠️ How?
- Integrated @ngx-translate and @angular/localize into the Angular application to manage translations.
- Enhanced StorageService to handle language preferences.
- Created and organized translation files (.json for frontend and .properties for backend) for all supported languages.
- Implemented language selection functionality on the frontend.
- Ensured that every backend response is localized based on the Accept-Language header sent by the frontend.
- Refactored the backend Message class into a more maintainable enum structure with translation keys.

## ✅ Testing?
1. In the top navigation bar, there is a language selection field, defaulting to English.
2. Select any language. The language of the page should change to the one chosen from the list.
3. Log in using one of the test accounts, e.g., with the user account (username: `user@example.com`, password: `userpass`), then place an order or reserve any book.
4. Go to `Profile` -> `Notifications`. The notification received from the server should be translated into the language you selected.
5. Go to `Actions`. After changing the language, the actions should be recorded in the selected language.

## 📸 Screenshots
![obraz](https://github.com/lukaszgardecki/library-app/assets/105553857/ca38155e-aad9-4cfa-9940-3850f16a2020)
![obraz](https://github.com/lukaszgardecki/library-app/assets/105553857/3b6f7ccf-5c3a-43ef-9bd0-7279321acabd)
![obraz](https://github.com/lukaszgardecki/library-app/assets/105553857/0aa418be-31f4-40f9-8809-9e377d243920)
![obraz](https://github.com/lukaszgardecki/library-app/assets/105553857/30121df6-86aa-4d3a-a24a-bc6b963ed031)
![obraz](https://github.com/lukaszgardecki/library-app/assets/105553857/01ab54a2-c3d3-4073-897f-ec1d73022016)
![obraz](https://github.com/lukaszgardecki/library-app/assets/105553857/dae636d9-b015-4ffd-95bf-ece6b8f77d9c)
